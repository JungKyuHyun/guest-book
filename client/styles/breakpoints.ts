/**
 * @see https://material-ui.com/customization/breakpoints/
 */
export const BREAK_POINTS = {
  sm: 600,
  md: 960,
  lg: 1280,
  xl: 1920,
} as const;
