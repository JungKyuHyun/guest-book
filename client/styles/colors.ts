/**
 * @see https://material.io/design/color/the-color-system.html#tools-for-picking-colors
 */
export const colors = {
  // red
  red50: '#ffebee',
  red100: '#ffcdd2',
  red200: '#ef9a9a',
  red300: '#e57373',
  red400: '#ef5350',
  red500: '#f44336',
  red600: '#e53935',
  red700: '#d32f2f',
  red800: '#c62828',
  red900: '#b71c1c',
  redA100: '#ff8a80',
  redA200: '#ff5252',
  redA400: '#ff1744',
  redA700: '#d50000',

  // pink
  pink50: '#fce4ec',
  pink100: '#f8bbd0',
  pink200: '#f48fb1',
  pink300: '#f06292',
  pink400: '#ec407a',
  pink500: '#e91e63',
  pink600: '#d81b60',
  pink700: '#c2185b',
  pink800: '#ad1457',
  pink900: '#880e4f',
  pinkA100: '#ff80ab',
  pinkA200: '#ff4081',
  pinkA400: '#f50057',
  pinkA700: '#c51162',

  // purple
  purple50: '#f3e5f5',
  purple100: '#e1bee7',
  purple200: '#ce93d8',
  purple300: '#ba68c8',
  purple400: '#ab47bc',
  purple500: '#9c27b0',
  purple600: '#8e24aa',
  purple700: '#7b1fa2',
  purple800: '#6a1b9a',
  purple900: '#4a148c',
  purpleA100: '#ea80fc',
  purpleA200: '#e040fb',
  purpleA400: '#d500f9',
  purpleA700: '#aa00ff',

  // deeppurple
  deeppurple50: '#ede7f6',
  deeppurple100: '#d1c4e9',
  deeppurple200: '#b39ddb',
  deeppurple300: '#9575cd',
  deeppurple400: '#7e57c2',
  deeppurple500: '#673ab7',
  deeppurple600: '#5e35b1',
  deeppurple700: '#512da8',
  deeppurple800: '#4527a0',
  deeppurple900: '#311b92',
  deeppurpleA100: '#b388ff',
  deeppurpleA200: '#7c4dff',
  deeppurpleA400: '#651fff',
  deeppurpleA700: '#6200ea',

  // indigo
  indigo50: '#e8eaf6',
  indigo100: '#c5cae9',
  indigo200: '#9fa8da',
  indigo300: '#7986cb',
  indigo400: '#5c6bc0',
  indigo500: '#3f51b5',
  indigo600: '#3949ab',
  indigo700: '#303f9f',
  indigo800: '#283593',
  indigo900: '#1a237e',
  indigoA100: '#8c9eff',
  indigoA200: '#536dfe',
  indigoA400: '#3d5afe',
  indigoA700: '#304ffe',

  // blue
  blue50: '#e3f2fd',
  blue100: '#bbdefb',
  blue200: '#90caf9',
  blue300: '#64b5f6',
  blue400: '#42a5f5',
  blue500: '#2196f3',
  blue600: '#1e88e5',
  blue700: '#1976d2',
  blue800: '#1565c0',
  blue900: '#0d47a1',
  blueA100: '#82b1ff',
  blueA200: '#448aff',
  blueA400: '#2979ff',
  blueA700: '#2962ff',

  // lightblue
  lightblue50: '#e1f5fe',
  lightblue100: '#b3e5fc',
  lightblue200: '#81d4fa',
  lightblue300: '#4fc3f7',
  lightblue400: '#29b6f6',
  lightblue500: '#03a9f4',
  lightblue600: '#039be5',
  lightblue700: '#0288d1',
  lightblue800: '#0277bd',
  lightblue900: '#01579b',
  lightblueA100: '#80d8ff',
  lightblueA200: '#40c4ff',
  lightblueA400: '#00b0ff',
  lightblueA700: '#0091ea',

  // cyan
  cyan50: '#e0f7fa',
  cyan100: '#b2ebf2',
  cyan200: '#80deea',
  cyan300: '#4dd0e1',
  cyan400: '#26c6da',
  cyan500: '#00bcd4',
  cyan600: '#00acc1',
  cyan700: '#0097a7',
  cyan800: '#00838f',
  cyan900: '#006064',
  cyanA100: '#84ffff',
  cyanA200: '#18ffff',
  cyanA400: '#00e5ff',
  cyanA700: '#00b8d4',

  // teal
  teal50: '#e0f2f1',
  teal100: '#b2dfdb',
  teal200: '#80cbc4',
  teal300: '#4db6ac',
  teal400: '#26a69a',
  teal500: '#009688',
  teal600: '#00897b',
  teal700: '#00796b',
  teal800: '#00695c',
  teal900: '#004d40',
  tealA100: '#a7ffeb',
  tealA200: '#64ffda',
  tealA400: '#1de9b6',
  tealA700: '#00bfa5',

  // green
  green50: '#e8f5e9',
  green100: '#c8e6c9',
  green200: '#a5d6a7',
  green300: '#81c784',
  green400: '#66bb6a',
  green500: '#4caf50',
  green600: '#43a047',
  green700: '#388e3c',
  green800: '#2e7d32',
  green900: '#1b5e20',
  greenA100: '#b9f6ca',
  greenA200: '#69f0ae',
  greenA400: '#00e676',
  greenA700: '#00c853',

  // lightgreen
  lightgreen50: '#f1f8e9',
  lightgreen100: '#dcedc8',
  lightgreen200: '#c5e1a5',
  lightgreen300: '#aed581',
  lightgreen400: '#9ccc65',
  lightgreen500: '#8bc34a',
  lightgreen600: '#7cb342',
  lightgreen700: '#689f38',
  lightgreen800: '#558b2f',
  lightgreen900: '#33691e',
  lightgreenA100: '#ccff90',
  lightgreenA200: '#b2ff59',
  lightgreenA400: '#76ff03',
  lightgreenA700: '#64dd17',

  // lime
  lime50: '#f9fbe7',
  lime100: '#f0f4c3',
  lime200: '#e6ee9c',
  lime300: '#dce775',
  lime400: '#d4e157',
  lime500: '#cddc39',
  lime600: '#c0ca33',
  lime700: '#afb42b',
  lime800: '#9e9d24',
  lime900: '#827717',
  limeA100: '#f4ff81',
  limeA200: '#eeff41',
  limeA400: '#c6ff00',
  limeA700: '#aeea00',

  // yellow
  yellow50: '#fffde7',
  yellow100: '#fff9c4',
  yellow200: '#fff59d',
  yellow300: '#fff176',
  yellow400: '#ffee58',
  yellow500: '#ffeb3b',
  yellow600: '#fdd835',
  yellow700: '#fbc02d',
  yellow800: '#f9a825',
  yellow900: '#f57f17',
  yellowA100: '#ffff8d',
  yellowA200: '#ffff00',
  yellowA400: '#ffea00',
  yellowA700: '#ffd600',

  // amber
  amber50: '#fff8e1',
  amber100: '#ffecb3',
  amber200: '#ffe082',
  amber300: '#ffd54f',
  amber400: '#ffca28',
  amber500: '#ffc107',
  amber600: '#ffb300',
  amber700: '#ffa000',
  amber800: '#ff8f00',
  amber900: '#ff6f00',
  amberA100: '#ffe57f',
  amberA200: '#ffd740',
  amberA400: '#ffc400',
  amberA700: '#ffab00',

  // orange
  orange50: '#fff3e0',
  orange100: '#ffe0b2',
  orange200: '#ffcc80',
  orange300: '#ffb74d',
  orange400: '#ffa726',
  orange500: '#ff9800',
  orange600: '#fb8c00',
  orange700: '#f57c00',
  orange800: '#ef6c00',
  orange900: '#e65100',
  orangeA100: '#ffd180',
  orangeA200: '#ffab40',
  orangeA400: '#ff9100',
  orangeA700: '#ff6d00',

  // deeporange
  deeporange50: '#fbe9e7',
  deeporange100: '#ffccbc',
  deeporange200: '#ffab91',
  deeporange300: '#ff8a65',
  deeporange400: '#ff7043',
  deeporange500: '#ff5722',
  deeporange600: '#f4511e',
  deeporange700: '#e64a19',
  deeporange800: '#d84315',
  deeporange900: '#bf360c',
  deeporangeA100: '#ff9e80',
  deeporangeA200: '#ff6e40',
  deeporangeA400: '#ff3d00',
  deeporangeA700: '#dd2c00',

  // brown
  brown50: '#efebe9',
  brown100: '#d7ccc8',
  brown200: '#bcaaa4',
  brown300: '#a1887f',
  brown400: '#8d6e63',
  brown500: '#795548',
  brown600: '#6d4c41',
  brown700: '#5d4037',
  brown800: '#4e342e',
  brown900: '#3e2723',

  // grey
  grey50: '#fafafa',
  grey100: '#f5f5f5',
  grey200: '#eeeeee',
  grey300: '#e0e0e0',
  grey400: '#bdbdbd',
  grey500: '#9e9e9e',
  grey600: '#757575',
  grey700: '#616161',
  grey800: '#424242',
  grey900: '#212121',

  // bluegrey
  bluegrey50: '#eceff1',
  bluegrey100: '#cfd8dc',
  bluegrey200: '#b0bec5',
  bluegrey300: '#90a4ae',
  bluegrey400: '#78909c',
  bluegrey500: '#607d8b',
  bluegrey600: '#546e7a',
  bluegrey700: '#455a64',
  bluegrey800: '#37474f',
  bluegrey900: '#263238',

  white: '#fff',
  black: '#000',
  transparent: 'transparent',
} as const;

export type Colors = keyof typeof colors;
