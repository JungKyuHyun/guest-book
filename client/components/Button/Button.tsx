import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { ButtonHTMLAttributes, ReactNode } from 'react';
import { boxShadows } from 'styles/boxshadows';
import {
  ButtonVariant,
  ButtonSize,
  SIZE_OF_STYLES,
  VARIANT_OF_STYLES,
} from './constants';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  isFill?: boolean;
  children?: ReactNode;
}

export function Button({
  variant = 'secondary',
  type = 'button',
  size = 'medium',
  isFill = false,
  disabled = false,
  children,
  ...props
}: ButtonProps) {
  return (
    <StyledButton
      type={type}
      size={size}
      variant={disabled ? 'disabled' : variant}
      disabled={disabled}
      isFill={isFill}
      {...props}
    >
      {children}
    </StyledButton>
  );
}

const StyledButton = styled.button<
  Required<Pick<ButtonProps, 'size' | 'variant' | 'isFill'>>
>`
  appearance: none;
  user-select: none;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  white-space: nowrap;
  cursor: pointer;
  border: 0;
  border-radius: 4px;
  line-height: 1.75;
  min-width: 64px;

  box-shadow: ${boxShadows.dp1};

  ${({ isFill }) =>
    isFill &&
    css`
      width: 100%;
    `};
  ${({ size }) => SIZE_OF_STYLES[size]};
  ${({ variant }) => VARIANT_OF_STYLES[variant]};

  &:disabled {
    box-shadow: none;
    pointer-events: none;
  }

  &:focus {
    outline: none;
  }
`;
