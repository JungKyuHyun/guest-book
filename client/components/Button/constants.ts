import { css, SerializedStyles } from '@emotion/react';
import { colors } from 'styles/colors';

const BUTTON_SIZE = {
  xSmall: 'xSmall',
  small: 'small',
  medium: 'medium',
  large: 'large',
} as const;

export type ButtonSize = keyof typeof BUTTON_SIZE;

const VARIANT = {
  primary: 'primary',
  secondary: 'secondary',
  dark: 'dark',
  danger: 'danger',
  disabled: 'disabled',
  text: 'text',
} as const;

export type ButtonVariant = keyof typeof VARIANT;

export const SIZE_OF_STYLES: { [key in ButtonSize]: SerializedStyles } = {
  [BUTTON_SIZE.large]: css`
    padding: 8px 22px;
    font-weight: 700;
    font-size: 15px;
    min-height: 48px;
  `,
  [BUTTON_SIZE.medium]: css`
    padding: 6px 16px;
    font-weight: 500;
    font-size: 14px;
    min-height: 40px;
  `,
  [BUTTON_SIZE.small]: css`
    padding: 4px 10px;
    font-weight: 500;
    font-size: 13px;
    min-height: 36px;
  `,
  [BUTTON_SIZE.xSmall]: css`
    padding: 4px;
    font-weight: 500;
    font-size: 13px;
    min-height: 30px;
    min-width: 48px;
  `,
};

export const VARIANT_OF_STYLES: { [key in ButtonVariant]: SerializedStyles } = {
  [VARIANT.primary]: css`
    background: ${colors.blue500};
    color: ${colors.black};

    &:hover {
      background: ${colors.blue600};
    }
  `,
  [VARIANT.secondary]: css`
    background: ${colors.grey300};
    color: ${colors.black};

    &:hover {
      background: ${colors.grey400};
    }
  `,
  [VARIANT.dark]: css`
    background: ${colors.black};
    color: ${colors.grey50};

    &:hover {
      background: ${colors.grey900};
    }
  `,
  [VARIANT.danger]: css`
    background: ${colors.red400};
    color: ${colors.grey50};

    &:hover {
      background: ${colors.red500};
    }
  `,
  [VARIANT.disabled]: css`
    background: ${colors.grey500};
    color: ${colors.grey300};
  `,
  [VARIANT.text]: css`
    background: transparent;
    color: ${colors.black};
    box-shadow: none;

    &:hover {
      background: ${colors.grey100};
    }
  `,
};
