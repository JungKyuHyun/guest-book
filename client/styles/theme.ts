import { boxShadows } from './boxshadows';
import { colors } from './colors';

export const theme = {
  color: colors,
  boxShadow: boxShadows,
};

export type Theme = typeof theme;
