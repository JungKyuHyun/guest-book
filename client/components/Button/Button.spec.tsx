import { Button } from './Button';
import { cleanup, fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { matchers } from '@emotion/jest';
import { colors } from 'styles/colors';
import { boxShadows } from 'styles/boxshadows';

expect.extend(matchers);

describe('<Button />', () => {
  afterEach(cleanup);

  it('match snapshot', () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    render(<Button />);
  });

  it('When passing children, draw it', () => {
    // arrange
    const childrenString = '버튼';
    const { getByText } = render(<Button>{childrenString}</Button>);

    // assert
    expect(getByText(childrenString)).toBeTruthy();
  });

  it('should call onClick prop when clicked', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick}>Click Me</Button>
    );
    fireEvent.click(getByText(/click me/i));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('should not call onClick prop when the "disabled" button clicked', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick} disabled>
        Click Me
      </Button>
    );
    fireEvent.click(getByText(/click me/i));
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  it('should render a button but no others', () => {
    const tree = renderer.create(<Button />).toJSON();
    // secondary style
    expect(tree).toHaveStyleRule('color', colors.black);
    expect(tree).toHaveStyleRule('background', colors.grey300);
    expect(tree).toHaveStyleRule('background', colors.grey400, {
      target: ':hover',
    });

    // medium size
    expect(tree).toHaveStyleRule('padding', '6px 16px');
    expect(tree).toHaveStyleRule('font-weight', '500');
    expect(tree).toHaveStyleRule('font-size', '14px');
    expect(tree).toHaveStyleRule('min-height', '40px');

    // common style
    expect(tree).toHaveStyleRule('box-shadow', boxShadows.dp1);
    expect(tree).toHaveStyleRule('line-height', '1.75');
    expect(tree).toHaveStyleRule('border-radius', '4px');
    expect(tree).toHaveStyleRule('min-width', '64px');
    expect(tree).toHaveStyleRule('cursor', 'pointer');
    expect(tree).toHaveStyleRule('border', '0');
    expect(tree).toHaveStyleRule('white-space', 'nowrap');
    expect(tree).toHaveStyleRule('appearance', 'none');
    expect(tree).toHaveStyleRule('user-select', 'none');
    expect(tree).toHaveStyleRule('display', 'inline-flex');
    expect(tree).toHaveStyleRule('justify-content', 'center');
    expect(tree).toHaveStyleRule('align-items', 'center');
    expect(tree).toHaveStyleRule('white-space', 'nowrap');
    expect(tree).toHaveStyleRule('box-shadow', 'none', {
      target: ':disabled',
    });
    expect(tree).toHaveStyleRule('pointer-events', 'none', {
      target: ':disabled',
    });
    expect(tree).toHaveStyleRule('outline', 'none', {
      target: ':focus',
    });
  });

  it('should render a primary button', () => {
    const tree = renderer.create(<Button variant='primary' />).toJSON();
    expect(tree).toHaveStyleRule('color', colors.black);
    expect(tree).toHaveStyleRule('background', colors.blue500);
    expect(tree).toHaveStyleRule('background', colors.blue600, {
      target: ':hover',
    });
  });

  it('should render a secondary', () => {
    const tree = renderer.create(<Button variant='secondary' />).toJSON();
    expect(tree).toHaveStyleRule('color', colors.black);
    expect(tree).toHaveStyleRule('background', colors.grey300);
    expect(tree).toHaveStyleRule('background', colors.grey400, {
      target: ':hover',
    });
  });

  it('should render a dark button', () => {
    const tree = renderer.create(<Button variant='dark' />).toJSON();
    expect(tree).toHaveStyleRule('color', colors.grey50);
    expect(tree).toHaveStyleRule('background', colors.black);
    expect(tree).toHaveStyleRule('background', colors.grey900, {
      target: ':hover',
    });
  });

  it('should render a disabled button', () => {
    const tree = renderer.create(<Button variant='disabled' />).toJSON();
    expect(tree).toHaveStyleRule('color', colors.grey300);
    expect(tree).toHaveStyleRule('background', colors.grey500);
  });

  it('should render a text button', () => {
    const tree = renderer.create(<Button variant='text' />).toJSON();
    expect(tree).toHaveStyleRule('color', colors.black);
    expect(tree).toHaveStyleRule('background', 'transparent');
    expect(tree).toHaveStyleRule('box-shadow', 'none');
    expect(tree).toHaveStyleRule('background', colors.grey100, {
      target: ':hover',
    });
  });

  it('should render a xSmall button', () => {
    const tree = renderer.create(<Button size='xSmall' />).toJSON();
    expect(tree).toHaveStyleRule('padding', '4px');
    expect(tree).toHaveStyleRule('font-weight', '500');
    expect(tree).toHaveStyleRule('font-size', '13px');
    expect(tree).toHaveStyleRule('min-height', '30px');
    expect(tree).toHaveStyleRule('min-width', '48px');
  });

  it('should render a small button', () => {
    const tree = renderer.create(<Button size='small' />).toJSON();
    expect(tree).toHaveStyleRule('padding', '4px 10px');
    expect(tree).toHaveStyleRule('font-weight', '500');
    expect(tree).toHaveStyleRule('font-size', '13px');
    expect(tree).toHaveStyleRule('min-height', '36px');
  });

  it('should render a medium button', () => {
    const tree = renderer.create(<Button size='medium' />).toJSON();
    expect(tree).toHaveStyleRule('padding', '6px 16px');
    expect(tree).toHaveStyleRule('font-weight', '500');
    expect(tree).toHaveStyleRule('font-size', '14px');
    expect(tree).toHaveStyleRule('min-height', '40px');
  });

  it('should render a large button', () => {
    const tree = renderer.create(<Button size='large' />).toJSON();
    expect(tree).toHaveStyleRule('padding', '8px 22px');
    expect(tree).toHaveStyleRule('font-weight', '700');
    expect(tree).toHaveStyleRule('font-size', '15px');
    expect(tree).toHaveStyleRule('min-height', '48px');
  });

  it('should render a fill button', () => {
    const tree = renderer.create(<Button isFill />).toJSON();
    expect(tree).toHaveStyleRule('width', '100%');
  });
});
