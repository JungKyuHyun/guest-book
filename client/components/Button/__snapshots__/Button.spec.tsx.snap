// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> match snapshot 1`] = `
<button
  className="css-1tfjm0-StyledButton-SIZE_OF_STYLES-VARIANT_OF_STYLES e1wvpnns0"
  disabled={false}
  size="medium"
  type="button"
/>
`;
